src/analyze_fs.c

```diff
@@ -351,12 +351,23 @@ static void iterate_directory(struct manifest *manifest, char *pathprefix,
 
 		string_or_die(&fullname, "%s/%s", fullpath, entry->d_name);
 		populate_file_struct(file, fullname);
-		free(fullname);
 
 		if (entry->d_type == DT_DIR) {
 			iterate_directory(manifest, pathprefix, file->filename, do_hash);
+		} else if (entry->d_type == DT_UNKNOWN) { /* fall back to lstat() */
+			struct stat sb;
```

Comment:
By referring to `get_hash()` I wanted to say that in the current code `populate_file_struct()` is called twice for every file: the first time it's called sequentially and then it's called in the parallel threads again. Which is probably not what you want. AFAIU this
```
if (entry->d_type == DT_DIR) {
```
check is an optimization postponing `lstat()` calls to the parallel threads.

So there are two solutions:

1. use `if (S_ISDIR(file.st_mode)) {`, remove the new call to `lstat()` and remove `populate_file_struct()` from `get_hash()`. This is easier to understand, but suboptimal since `lstat()` is called sequentially for all files.
2. drop the first call to `populate_file_struct()`. In this case the second `lstat()` is called only as a fallback for rare cases where `DT_DIR` is not supported.